### PR TITLE
Inm/create gradle task to run pilates

### PR DESCRIPTION
**Pull request by irishwithaxe**
1. created `runPilates` aws lambda function 
2. changed `updateBeamAMI` aws lambda function to further updates of `runPilates` AMI's
3. created gradle task to run `runPilates` function
4. added documentation entry with short description of gradle task and lambda function
5. created src/main/bash/runPilates-examples.sh file with examples of the gradle task calls
6. small fix of gradle task `deploy`
7. all AWS tasks moved into build.gradle in aws folder
8. gradle configured to be able to run all tasks from root folder (including aws tasks)
9. gradle.properties from aws folder removed because it is redundant

corresponding pilates changes - https://github.com/LBNL-UCB-STI/PILATES/pull/1

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2228)

### CODE REVIEW

File: aws/src/main/python/runPilates/lambda_function.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+                'Key': 'Name',
+                'Value': instance_name
+            }]
+        }])
+    return res['Instances'][0]['InstanceId']
+
+
+def get_dns(instance_id):
+    host = None
+    while host is None:
```
Comment (JustinPihony) [inline aws/src/main/python/runPilates/lambda_function.py]: Maybe this should have some form of max tries in case the instance failed for some reason - otherwise this will spin forever. Or an exponential backoff?

File: aws/src/main/python/runPilates/lambda_function.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        instances = ec2.describe_instances(InstanceIds=[instance_id])
+        for r in instances['Reservations']:
+            for i in r['Instances']:
+                dns = i['PublicDnsName']
+                if dns != '':
+                    host = dns
+    return host
+
+
+def check_instance_id(instance_ids):
```
Comment (JustinPihony) [inline aws/src/main/python/runPilates/lambda_function.py]: This is not used and seems able to be deleted

File: aws/src/main/python/runPilates/lambda_function.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    for reservation in ec2.describe_instances()['Reservations']:
+        for instance in reservation['Instances']:
+            if instance['InstanceId'] in instance_ids:
+                instance_ids.remove(instance['InstanceId'])
+    return instance_ids
+
+
+def start_instance(instance_ids):
```
Comment (JustinPihony) [inline aws/src/main/python/runPilates/lambda_function.py]: This is not used and seems able to be deleted

File: aws/src/main/python/runPilates/lambda_function.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    return ec2.start_instances(InstanceIds=instance_ids)
+
+
+def stop_instance(instance_ids):
```
Comment (JustinPihony) [inline aws/src/main/python/runPilates/lambda_function.py]: This is not used and seems able to be deleted

File: aws/src/main/python/runPilates/lambda_function.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    return ec2.stop_instances(InstanceIds=instance_ids)
+
+
+def terminate_instance(instance_ids):
```
Comment (JustinPihony) [inline aws/src/main/python/runPilates/lambda_function.py]: This is not used and seems able to be deleted

File: build.gradle [hunk 6]
```diff
@@ -384,6 +384,73 @@ def getCurrentGitBranch() {
     return gitBranch
 }
 
+// gradle deployPilates -PrunName=inm/testRunPilates -PstartYear=2010 -PcountOfYears=15 -PbeamItLen=5 -PurbansimItLen=5 -PbeamBranch=production-sfbay-develop -PbeamConfigs=production/sfbay/smart/smart-baseline-pilates-small-test.conf -PinitialS3UrbansimInput=//inm-test-run-pilates/initial_input/ -PpilatesImageName=irishwithaxe/pilates-inm -PpilatesScenarioName=pilatesScenarioName1 -PoutputBucketBasePath=//inm-test-run-pilates/pilates-outputs -PinitialSkimsPath=//inm-test-run-pilates/pilates-outputs/pilatesScenarioName1_/2015/beam/sfbay-smart-base-pilates-small__2019-10-08_20-54-59/ITERS/it.0/0.skims.csv.gz
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: I am agree, going to remove those comments

File: aws/src/main/python/runPilates/lambda_function.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    return ec2.terminate_instances(InstanceIds=instance_ids)
+
+
+def lambda_handler(event, context):
+    titled = event.get('title', 'hostname-test')
+    if titled is None:
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: agree, let it be None by default

File: aws/src/main/python/runPilates/lambda_function.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        return "Unable to start the run, runName is required. Please restart with appropriate runName."
+
+    initial_urbansim_input = event.get('initial_urbansim_input', None)
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: oh, this is even better

File: aws/src/main/python/runPilates/lambda_function.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    if initial_urbansim_input is None:
+        return "Unable to start the run, initialS3UrbansimInput is required. Please restart with appropriate arguments."
+
+    initial_urbansim_output = event.get('initial_urbansim_output', None)
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: +

File: aws/src/main/python/runPilates/lambda_function.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+PILATES_IMAGE_VERSION_DEFAULT = 'latest'
+PILATES_IMAGE_NAME_DEFAULT = 'pilates'
+PILATES_SCENARIO_NAME_DEFAULT = 'pilates'
+OUTPUT_BUCKET_BASE_PATH_DEFAULT = '//pilates-outputs'
+IN_YEAR_OUTPUT_DEFAULT = 'off'
+START_YEAR_DEFAULT = '2010'
+COUNT_OF_YEARS_DEFAULT = '30'
+BEAM_IT_LEN_DEFAULT = '5'
+URBANSIM_IT_LEN_DEFAULT = '5'
+BRANCH_DEFAULT = 'master'
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: I think it even should not have default value, going to remove it

File: build.gradle [hunk 11]
```diff
@@ -384,6 +384,73 @@ def getCurrentGitBranch() {
...
+        if (!project.hasProperty('initialS3UrbansimInput')) {
+            throw new GradleException('`initialS3UrbansimInput` argument required.' + runExample)
+        }
+    }
+
+    def pload = """{
+  "title": "${project.findProperty('runName')}",
+  "branch": "${project.findProperty('beamBranch') ?: getCurrentGitBranch()}",
+  "commit": "${project.findProperty('beamCommit') ?: 'HEAD'}",
+  "start_year": "${project.findProperty('startYear')}",
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: good point, will test that in corner cases

File: aws/src/main/python/runPilates/lambda_function.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+        shutdown_behaviour = os.environ['SHUTDOWN_BEHAVIOUR']
+
+    if volume_size < 64 or volume_size > 256:
+        volume_size = 64
+
+    selected_script = RUN_PILATES_SCRIPT
+
+    params = configs
+    if batch == TRUE:
+        params = [params.replace(',', ' ')]
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: yes, this was really strange things to read.
now it is gone for good because we do not need batch runs in pilates.
but it is still in beam deploy lambda function

File: aws/src/main/python/runPilates/lambda_function.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    volume_size = event.get('storage_size', 64)
+    shutdown_wait = event.get('shutdown_wait', SHUTDOWN_DEFAULT)
+    region = event.get('region', os.environ['REGION'])
+    shutdown_behaviour = event.get('shutdown_behaviour', os.environ['SHUTDOWN_BEHAVIOUR'])
+    sigopt_client_id = event.get('sigopt_client_id', os.environ['SIGOPT_CLIENT_ID'])
+    sigopt_dev_id = event.get('sigopt_dev_id', os.environ['SIGOPT_DEV_ID'])
+    end_script = event.get('end_script', END_SCRIPT_DEFAULT)
+
+    if initial_urbansim_output is None and initial_skims_path is '':
+        return "Unable to start run, initialS3UrbansimOutput (initial beam data) or initialSkimsPath (initial skims file) should be specified."
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: yes, it is really OR. someone should specify one or another.

File: aws/src/main/python/runPilates/lambda_function.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+        for arg in params:
+            uid = str(uuid.uuid4())[:8]
+            run_name = titled
+            if len(params) > 1:
+                run_name += "-" + repr(run_num)
+
+            script = initscript.replace('$RUN_SCRIPT', selected_script).replace('$REGION', region) \
+                .replace('$PREPARE_URBANSIM_INPUT_SCRIPT', PREPARE_URBANSIM_INPUT_SCRIPT) \
+                .replace('$PREPARE_URBANSIM_OUTPUT_SCRIPT', prepareUrbansimOutScript) \
+                .replace('$S3_REGION', os.environ['REGION']) \
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: done

File: aws/src/main/python/runPilates/lambda_function.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+                .replace('$SLACK_HOOK_WITH_TOKEN', os.environ['SLACK_HOOK_WITH_TOKEN'])
+
+            instance_id = deploy(script, instance_type, region.replace("-", "_") + '_', shutdown_behaviour, run_name,
+                                 volume_size)
+            host = get_dns(instance_id)
+            txt = txt + 'Started batch: {batch} with run name: {titled} for branch/commit {branch}/{commit} at host {dns} (InstanceID: {instance_id}). '.format(
+                branch=branch, titled=run_name, commit=commit_id, dns=host, batch=uid, instance_id=instance_id)
+            run_num += 1
+    else:
+        txt = 'Unable to start bach for branch/commit {branch}/{commit}. '.format(branch=branch, commit=commit_id)
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: batch functionality was removed completely

File: build.gradle [hunk 16]
```diff
@@ -384,6 +384,73 @@ def getCurrentGitBranch() {
...
+  "urbansim_it_len": "${project.findProperty('urbansimItLen')}",
+  "initial_urbansim_input": "${project.findProperty('initialS3UrbansimInput')}",
+  "initial_urbansim_output": "${project.findProperty('initialS3UrbansimOutput')}",
+  "pilates_image_version": "${project.findProperty('pilatesImageVersion') ?: 'latest'}",
+  "pilates_image_name": "${project.findProperty('pilatesImageName') ?: 'beammodel/pilates'}",
+  "pilates_scenario_name": "${project.findProperty('pilatesScenarioName') ?: 'pilates'}",
+  "initial_skims_path": "${project.findProperty('initialSkimsPath') ?: ''}",
+  "output_bucket_base_path": "${project.findProperty('outputBucketBasePath') ?: '//pilates-outputs'}",
+  "in_year_output": "${project.findProperty('inYearOutput') ?: 'off'}",
+  "configs": "${project.findProperty('beamConfigs') ?: project.findProperty(getCurrentGitBranch() + '.configs')}",
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: i assume that was for batch functionality which i have removed now

File: build.gradle [hunk 17]
```diff
@@ -384,6 +384,73 @@ def getCurrentGitBranch() {
...
+  "max_ram": "${maxRAM ?: '140g'}",
+  "storage_size": ${project.findProperty('storageSize')},
+  "batch": "$beamBatch",
+  "instance_type": "${project.findProperty('instanceType') ?: defaultInstanceType}",
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: because instanceType is specified in gradle.properties and .. looks like it is `double default` value (=
if it is not specified as parameter it will be read from gradle.properties and only if there no `instanceType` variable even in gradle.properties then `defaultInstanceType` will be used.

by the way, in my copy of gradle.properties there is a `defaultInstanceType=t2.small` line

File: build.gradle [hunk 18]
```diff
@@ -384,6 +384,73 @@ def getCurrentGitBranch() {
...
+  "region": "$region",
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: agree.
I will change them to look like other parameters (with value firstly taken from project property)

File: aws/src/main/python/runPilates/lambda_function.py [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+    batch = event.get('batch', TRUE)
+    max_ram = event.get('max_ram', MAXRAM_DEFAULT)
+    instance_type = event.get('instance_type', os.environ['INSTANCE_TYPE'])
+    volume_size = event.get('storage_size', 64)
+    shutdown_wait = event.get('shutdown_wait', SHUTDOWN_DEFAULT)
+    region = event.get('region', os.environ['REGION'])
+    shutdown_behaviour = event.get('shutdown_behaviour', os.environ['SHUTDOWN_BEHAVIOUR'])
+    sigopt_client_id = event.get('sigopt_client_id', os.environ['SIGOPT_CLIENT_ID'])
+    sigopt_dev_id = event.get('sigopt_dev_id', os.environ['SIGOPT_DEV_ID'])
+    end_script = event.get('end_script', END_SCRIPT_DEFAULT)
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: agree, I have already removed that `end_script`

File: aws/src/main/python/runPilates/lambda_function.py (hunk 5 above)
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: done

File: aws/src/main/python/runPilates/lambda_function.py (hunk 4 above)
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: done

File: aws/src/main/python/runPilates/lambda_function.py (hunk 3 above)
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: done

File: aws/src/main/python/runPilates/lambda_function.py (hunk 2 above)
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: done

File: aws/src/main/python/runPilates/lambda_function.py [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+ec2 = None
+
+
+def init_ec2(region):
+    global ec2
+    ec2 = boto3.client('ec2', region_name=region)
+
+
+def validate(name):
+    return True
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: done

File: aws/src/main/python/runPilates/lambda_function.py [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+PILATES_SCENARIO_NAME_DEFAULT = 'pilates'
+OUTPUT_BUCKET_BASE_PATH_DEFAULT = '//pilates-outputs'
+IN_YEAR_OUTPUT_DEFAULT = 'off'
+START_YEAR_DEFAULT = '2010'
+COUNT_OF_YEARS_DEFAULT = '30'
+BEAM_IT_LEN_DEFAULT = '5'
+URBANSIM_IT_LEN_DEFAULT = '5'
+BRANCH_DEFAULT = 'master'
+COMMIT_DEFAULT = 'HEAD'
+MAXRAM_DEFAULT = '2g'
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: i will just remove those default values in lambda to be able to have only one source of default values

File: aws/src/main/python/runPilates/lambda_function.py (hunk 1 above)
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: good point, will add something like max_tries

File: aws/src/main/python/runPilates/lambda_function.py [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+        return "Unable to start run, initialS3UrbansimOutput (initial beam data) or initialSkimsPath (initial skims file) should be specified."
+
+    if instance_type not in instance_types:
+        return "Unable to start run, {instance_type} instance type not supported.".format(instance_type=instance_type)
+
+    if shutdown_behaviour not in shutdown_behaviours:
+        shutdown_behaviour = os.environ['SHUTDOWN_BEHAVIOUR']
+
+    if volume_size < 64 or volume_size > 256:
+        volume_size = 64
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply aws/src/main/python/runPilates/lambda_function.py]: sounds logical, will do that

File: build.gradle [hunk 23]
```diff
@@ -437,7 +437,7 @@ task deployPilates(type: AWSLambdaInvokeTask) {
   "storageSize": ${project.findProperty('storageSize')},
   "instanceType": "${project.findProperty('instanceType') ?: defaultInstanceType}",
   "dataRegion": "${project.findProperty('dataRegion') ?: region}",
-  "maxRam": "${project.findProperty('maxRAM') ?: 140g}",
+  "maxRam": "${project.findProperty('maxRAM')}",
```
[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: if there is no direct value, then gradle will try to read value from gradle.properties and if there will be no value too then gradle will return error like `Could not get unknown property 'maxRAM' for task`

[the remark this answers is not in the document]
Comment (irishwithaxe) [reply build.gradle]: and anyway, if gradle somehow will manage to sent wrong value (like `null`) then lambda function return error with text similar to `missing parameter: maxRAM`

**Comment by irishwithaxe** [comment]
Test!

**Comment by irishwithaxe** [comment]
Test!

**Comment by irishwithaxe** [comment]
Test!

**Comment by irishwithaxe** [comment]
Test!

**Comment by irishwithaxe** [comment]
Test!

**Comment by irishwithaxe** [comment]
Test!

**Comment by irishwithaxe** [comment]
Test!
